php_phongo.c

```diff
@@ -2841,18 +2841,379 @@ static bool phongo_manager_set_auto_encryption_opts(php_phongo_manager_t* manage
+
+			goto cleanup;
+		}
+
+		mongoc_client_encryption_opts_set_keyvault_namespace(opts, db_name, coll_name);
+		efree(db_name);
+		efree(coll_name);
+
+		if (pfree) {
+			efree(keyvault_namespace);
+		}
+	}
+
+	if (php_array_existsc(options, "kmsProviders")) {
+		zval*  kms_providers  = php_array_fetchc(options, "kmsProviders");
+		bson_t bson_providers = BSON_INITIALIZER;
+
+		if (Z_TYPE_P(kms_providers) != IS_ARRAY) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"kmsProviders\" encryption option to be an array");
+			goto cleanup;
+		}
+
+		php_phongo_zval_to_bson(kms_providers, PHONGO_BSON_NONE, &bson_providers, NULL TSRMLS_CC);
+		if (EG(exception)) {
+			goto cleanup;
+		}
+
+		mongoc_client_encryption_opts_set_kms_providers(opts, &bson_providers);
+		bson_destroy(&bson_providers);
+	}
+
+	return opts;
+
+cleanup:
+	if (opts) {
+		mongoc_client_encryption_opts_destroy(opts);
+	}
+
+	return NULL;
+} /* }}} */
+
+void phongo_clientencryption_init(php_phongo_clientencryption_t* clientencryption, mongoc_client_t* client, zval* options TSRMLS_DC) /* {{{ */
+{
+	mongoc_client_encryption_t*      ce;
+	mongoc_client_encryption_opts_t* opts;
+	bson_error_t                     error = { 0 };
+
+	opts = phongo_clientencryption_opts_from_zval(client, options TSRMLS_CC);
+
+	ce = mongoc_client_encryption_new(opts, &error);
+	if (!ce) {
+		phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
+
+		goto cleanup;
+	}
+
+	clientencryption->client_encryption = ce;
+
+cleanup:
+	if (opts) {
+		mongoc_client_encryption_opts_destroy(opts);
+	}
+} /* }}} */
+
+static mongoc_client_encryption_datakey_opts_t* phongo_clientencryption_datakey_opts_from_zval(zval* options TSRMLS_DC) /* {{{ */
+{
+	mongoc_client_encryption_datakey_opts_t* opts;
+
+	opts = mongoc_client_encryption_datakey_opts_new();
+
+	if (!options || Z_TYPE_P(options) != IS_ARRAY) {
+		return opts;
+	}
+
+	if (php_array_existsc(options, "keyAltNames")) {
+		zval*      zkeyaltnames = php_array_fetchc(options, "keyAltNames");
+		HashTable* ht_data;
+		uint32_t   keyaltnames_count;
+		char**     keyaltnames;
+		uint32_t   i      = 0;
+		bool       failed = false;
+
+		if (!zkeyaltnames || Z_TYPE_P(zkeyaltnames) != IS_ARRAY) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected keyAltNames to be array, %s given", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(zkeyaltnames));
+			goto cleanup;
+		}
+
+		ht_data           = HASH_OF(zkeyaltnames);
+		keyaltnames_count = ht_data ? zend_hash_num_elements(ht_data) : 0;
+		keyaltnames       = bson_malloc(sizeof(char*) * keyaltnames_count);
+
+#if PHP_VERSION_ID >= 70000
+		{
+			zval* keyaltname;
+
+			ZEND_HASH_FOREACH_VAL(ht_data, keyaltname)
+			{
+				if (i == keyaltnames_count) {
+					phongo_throw_exception(PHONGO_ERROR_LOGIC TSRMLS_CC, "Iterating over too many keyAltNames. Please file a bug report");
+					failed = true;
+				}
+
+				keyaltnames[i] = bson_strdup(Z_STRVAL_P(keyaltname));
```

Comment:
I think it'd make sense to do a type check on `keyaltname` at the start of the `ZEND_HASH_FOREACH_VAL` scope. If we find a non-string in the array, then we can raise a consistently formatted type exception.

In practice, this would probably be a non-associative array, but if you want to generate the most helpful exception message we may want to use `ZEND_HASH_FOREACH_KEY_VAL` so we can report the key. I believe both it makes the string and numeric key easily accessible. Take a look at `php_phongo_bson_state_parse_fieldpaths` for prior art if you go this route. Alternatively, I'm happy to punt on improving the error message to a future ticket after 1.7.0.

Reply:
Implemented type checks along with the proper error message depending on the array key. Also included tests for the behaviour.